utils/searxng.sh

```diff
@@ -488,9 +488,6 @@ EOF
     info_msg "install needed python packages"
     tee_stderr 0.1 <<EOF | sudo -H -u "${SERVICE_USER}" -i 2>&1 | prefix_stdout "$_service_prefix"
 pip install -U pip
-pip install -U setuptools
-pip install -U wheel
-pip install -U pyyaml
```

Comment:
AFAIK we can't remove these packages from the *pyenv*.  By example: in setup.py we import `setuptools` .. if we do not install an up-to-date `setuptools`  *pyenv* the `setuptools`  from OS is used and this might be an old version.  Same wtih wheel .. if we do not install those packages in the *pyenv* we relay on the versions of this packages in the OS. To be independent from what is installed on a OS we have to install theses components in the *pyenv*.

Reply:
[PEP 0517](https://peps.python.org/pep-0517/) to the rescue: see `pyproject.toml` which describes which dependencies to install to allow the searxng to be installed. In other words, `pip` takes into account `pyproject.toml` to provide the required dependencies when `setup.py` runs. The PEP 0517 exists since 2015 and the last update was in 2017.

[fasttext-predict](https://github.com/searxng/fasttext-predict) and [httpx](https://github.com/encode/httpx/) use `pyproject.toml`.

With this, SearXNG can be installed with pip:

<details>
<summary>pip install searxng@git+https://github.com/dalf/searxng@pyprojecttoml#egg=searxng</summary>


`@pyprojecttoml` refers to the branch

```sh
$ python -m venv venv
$ . ./venv/bin/activate
$ pip install -U pip
$ pip install searxng@git+https://github.com/dalf/searxng@pyprojecttoml#egg=searxng
Collecting searxng@ git+https://github.com/dalf/searxng@pyprojecttoml#egg=searxng
  Cloning https://github.com/dalf/searxng (to revision pyprojecttoml) to /tmp/pip-install-lpz7m__9/searxng_2fc49791cd1d49f898e6495754970264
  Running command git clone --filter=blob:none --quiet https://github.com/dalf/searxng /tmp/pip-install-lpz7m__9/searxng_2fc49791cd1d49f898e6495754970264
  Running command git checkout -b pyprojecttoml --track origin/pyprojecttoml
  Switched to a new branch 'pyprojecttoml'
  Branch 'pyprojecttoml' set up to track remote branch 'pyprojecttoml' from 'origin'.
  Resolved https://github.com/dalf/searxng to commit be9734ebee87604cf1c6ca2a7dcf9af36cf75dba
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting python-dateutil==2.8.2
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting jinja2==3.1.2
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
Collecting httpx[http2]==0.21.2
  Using cached httpx-0.21.2-py3-none-any.whl (83 kB)
Collecting uvloop==0.17.0
  Using cached uvloop-0.17.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.2 MB)
Collecting flask==2.2.2
  Using cached Flask-2.2.2-py3-none-any.whl (101 kB)
Collecting lxml==4.9.2
  Using cached lxml-4.9.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (7.1 MB)
Collecting pyyaml==6.0
  Using cached PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (661 kB)
Collecting babel==2.11.0
  Using cached Babel-2.11.0-py3-none-any.whl (9.5 MB)
Collecting typing-extensions==4.4.0
  Using cached typing_extensions-4.4.0-py3-none-any.whl (26 kB)
Collecting markdown-it-py==2.1.0
  Using cached markdown_it_py-2.1.0-py3-none-any.whl (84 kB)
Collecting redis==4.4.2
  Using cached redis-4.4.2-py3-none-any.whl (237 kB)
Collecting httpx-socks[asyncio]==0.7.2
  Using cached httpx_socks-0.7.2-py3-none-any.whl (12 kB)
Collecting setproctitle==1.3.2
  Using cached setproctitle-1.3.2-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (30 kB)
Collecting fasttext-predict==0.9.2.1
  Using cached fasttext_predict-0.9.2.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (115 kB)
Collecting flask-babel==2.0.0
  Using cached Flask_Babel-2.0.0-py3-none-any.whl (9.3 kB)
Collecting certifi==2022.12.7
  Using cached certifi-2022.12.7-py3-none-any.whl (155 kB)
Collecting pygments==2.14.0
  Using cached Pygments-2.14.0-py3-none-any.whl (1.1 MB)
Collecting Brotli==1.0.9
  Using cached Brotli-1.0.9-cp39-cp39-manylinux1_x86_64.whl (357 kB)
Collecting pytz>=2015.7
  Using cached pytz-2022.7.1-py2.py3-none-any.whl (499 kB)
Collecting importlib-metadata>=3.6.0
  Using cached importlib_metadata-6.0.0-py3-none-any.whl (21 kB)
Collecting itsdangerous>=2.0
  Using cached itsdangerous-2.1.2-py3-none-any.whl (15 kB)
Collecting click>=8.0
  Using cached click-8.1.3-py3-none-any.whl (96 kB)
Collecting Werkzeug>=2.2.2
  Using cached Werkzeug-2.2.2-py3-none-any.whl (232 kB)
Collecting httpcore<0.15.0,>=0.14.0
  Using cached httpcore-0.14.7-py3-none-any.whl (68 kB)
Collecting httpx<0.22.0,>=0.21.0
  Downloading httpx-0.21.3-py3-none-any.whl (83 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 83.8/83.8 KB 339.7 kB/s eta 0:00:00
Collecting python-socks>=2.0.0
  Using cached python_socks-2.1.1-py3-none-any.whl (49 kB)
Collecting async-timeout>=3.0.1
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting charset-normalizer
  Using cached charset_normalizer-3.0.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (198 kB)
Collecting sniffio
  Using cached sniffio-1.3.0-py3-none-any.whl (10 kB)
Collecting rfc3986[idna2008]<2,>=1.3
  Using cached rfc3986-1.5.0-py2.py3-none-any.whl (31 kB)
Collecting h2<5,>=3
  Using cached h2-4.1.0-py3-none-any.whl (57 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.1.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Collecting mdurl~=0.1
  Using cached mdurl-0.1.2-py3-none-any.whl (10.0 kB)
Collecting six>=1.5
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting hyperframe<7,>=6.0
  Using cached hyperframe-6.0.1-py3-none-any.whl (12 kB)
Collecting hpack<5,>=4.0
  Using cached hpack-4.0.0-py3-none-any.whl (32 kB)
Collecting anyio==3.*
  Using cached anyio-3.6.2-py3-none-any.whl (80 kB)
Collecting h11<0.13,>=0.11
  Using cached h11-0.12.0-py3-none-any.whl (54 kB)
Collecting idna>=2.8
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting zipp>=0.5
  Using cached zipp-3.11.0-py3-none-any.whl (6.6 kB)
Building wheels for collected packages: searxng
  Building wheel for searxng (pyproject.toml) ... done
  Created wheel for searxng: filename=searxng-2023.1.22+gbe9734eb-py3-none-any.whl size=4280311 sha256=6d97782f47b927f5eb32a8ae65be848179cc162cef92dd65f72ca4a9a34be822
  Stored in directory: /tmp/pip-ephem-wheel-cache-evhd08c4/wheels/fb/cf/7d/ff5416bd569581df3fc842d277520ee518657317a710d215cd
Successfully built searxng
Installing collected packages: rfc3986, pytz, python-socks, fasttext-predict, charset-normalizer, Brotli, zipp, uvloop, typing-extensions, sniffio, six, setproctitle, pyyaml, pygments, mdurl, MarkupSafe, lxml, itsdangerous, idna, hyperframe, hpack, h11, click, certifi, babel, async-timeout, Werkzeug, redis, python-dateutil, markdown-it-py, jinja2, importlib-metadata, h2, anyio, httpcore, flask, httpx, flask-babel, httpx-socks, searxng
Successfully installed Brotli-1.0.9 MarkupSafe-2.1.2 Werkzeug-2.2.2 anyio-3.6.2 async-timeout-4.0.2 babel-2.11.0 certifi-2022.12.7 charset-normalizer-3.0.1 click-8.1.3 fasttext-predict-0.9.2.1 flask-2.2.2 flask-babel-2.0.0 h11-0.12.0 h2-4.1.0 hpack-4.0.0 httpcore-0.14.7 httpx-0.21.2 httpx-socks-0.7.2 hyperframe-6.0.1 idna-3.4 importlib-metadata-6.0.0 itsdangerous-2.1.2 jinja2-3.1.2 lxml-4.9.2 markdown-it-py-2.1.0 mdurl-0.1.2 pygments-2.14.0 python-dateutil-2.8.2 python-socks-2.1.1 pytz-2022.7.1 pyyaml-6.0 redis-4.4.2 rfc3986-1.5.0 searxng-2023.1.22+gbe9734eb setproctitle-1.3.2 six-1.16.0 sniffio-1.3.0 typing-extensions-4.4.0 uvloop-0.17.0 zipp-3.11.0
```

To show everything is fine, you can start SearXNG with the `searxng` command (the app starts with Werkzeug, this is only to show it works):
```
$ SEARXNG_DEBUG=1 searxng
INFO    searx                         : load the default settings from /home/alexandre/code/tmp/venv/lib/python3.9/site-packages/searx/settings.yml
INFO    searx                         : max_request_timeout=None
ERROR   searx                         : Error while getting the version: fatal: not a git repository (or any of the parent directories): .git

ERROR   searx                         : Error while getting the git URL & branch: fatal: not a git repository (or any of the parent directories): .git

INFO    searx                         : version: 1.0.0
DEBUG   searx.webapp                  : static directory is /home/alexandre/code/tmp/venv/lib/python3.9/site-packages/searx/static
DEBUG   searx.webapp                  : templates directory is /home/alexandre/code/tmp/venv/lib/python3.9/site-packages/searx/templates
DEBUG   searx.locales                 : locales_initialize: /home/alexandre/code/tmp/venv/lib/python3.9/site-packages/searx/translations
DEBUG   searx.plugins                 : searx.plugins.ahmia_filter: loaded
DEBUG   searx.plugins                 : searx.plugins.autodetect_search_language: loaded
DEBUG   searx.plugins                 : searx.plugins.hash_plugin: loaded
DEBUG   searx.plugins                 : searx.plugins.hostname_replace: loaded
DEBUG   searx.plugins                 : searx.plugins.limiter: loaded
DEBUG   searx.plugins                 : searx.plugins.oa_doi_rewrite: loaded
DEBUG   searx.plugins                 : searx.plugins.search_on_category_select: loaded
...
DEBUG   searx.engines.soundcloud      : Initialized
DEBUG   searx.network.wolframalpha    : HTTP Request: GET https://www.wolframalpha.com/input/api/v1/code?ts=9999999999999999999 "HTTP/1.1 200 OK"
DEBUG   charset_normalizer            : Encoding detection: ascii is most likely the one.
DEBUG   searx.engines.wolframalpha    : Initialized
DEBUG   searx.engines.wikidata        : request time 0:00:01.070000
DEBUG   searx.engines.wikidata        : Initialized
```

</details>

Then `requirements.txt` contains the required dependencies for development, this include `-e .` to install the SearXNG package.

httpx does this, see:
* https://github.com/encode/httpx/blob/master/scripts/install
* https://github.com/encode/httpx/blob/59914c7690573d5f9b17bce4172405294cb08278/requirements.txt#L5

Reply:
> [PEP 0517](https://peps.python.org/pep-0517/) to the rescue: see `pyproject.toml` which describes which dependencies to install to allow the searxng to be installed. 

sounds promising .. I do not have experience with pip & toml .. do we need to install an up-to-date toml parser? .. I assume we get one when we do a `install -U pip` .. its included in pip right?

If it is included in pip, can we build the *pyenv* by Pip & toml? .. if so we can comsider to drop/replace the entire `searxng.install.pyenv()` bash function.

Reply:
> do we need to install an up-to-date toml parser?

it requires a not too old pip, nothing more: see `fasttext-predict`, we don't care about toml.

> can we build the pyenv by Pip & toml?

yes, at first sight, I see these updates:
* `pyenv.OK()` checks if `requirements.txt` and `setup.py` have changed (instead of `requirements*.txt`)
* merge `pyenv.install.OK` and `pyenv.OK` : if the `searx` package is not installed, then the virtualenv can be dropped (I'm not about this one, I'm lost between `manage` and `utils/lib.sh`).
* `pyenv.install()` installs everything with `pip -r requirements.txt` (no need to install `requirements.txt` and then `-e .[test]`)


```bash
pyenv.install() {

    if ! pyenv.OK; then
        py.clean > /dev/null
    fi
    if pyenv.install.OK > /dev/null; then
        return 0
    fi

    (   set -e
        pyenv
        build_msg PYENV "[install] pip install -r requirements.txt"
        "${PY_ENV_BIN}/python" -m pip install r requirements.txt
        buildenv
    )
    local exit_val=$?
    if [ ! $exit_val -eq 0 ]; then
        die 42 "error while pip install (${PY_ENV_BIN})"
    fi
}
```